File: src/main/java/com/wzh/intelligentcodeapp/controller/UserController.java
package com.wzh.intelligentcodeapp.controller;

import cn.hutool.core.bean.BeanUtil;
import cn.hutool.core.util.ObjectUtil;
import com.mybatisflex.core.paginate.Page;
import com.wzh.intelligentcodeapp.annotation.AuthCheck;
import com.wzh.intelligentcodeapp.common.BaseResponse;
import com.wzh.intelligentcodeapp.common.ResultUtils;
import com.wzh.intelligentcodeapp.exception.ErrorCode;
import com.wzh.intelligentcodeapp.exception.ThrowUtils;
import com.wzh.intelligentcodeapp.model.request.UserListRequest;
import com.wzh.intelligentcodeapp.model.request.UserLoginRequest;
import com.wzh.intelligentcodeapp.model.request.UserRegisterRequest;
import com.wzh.intelligentcodeapp.model.vo.LoginUserVO;
import jakarta.servlet.http.HttpServletRequest;
import org.springframework.web.bind.annotation.DeleteMapping;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.PutMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.beans.factory.annotation.Autowired;
import com.wzh.intelligentcodeapp.model.entity.User;
import com.wzh.intelligentcodeapp.service.UserService;
import org.springframework.web.bind.annotation.RestController;

import java.util.ArrayList;
import java.util.List;

/**
 * 用户 控制层。
 *
 * @author wzh
 * @since 2025-11-26
 */
@RestController
@RequestMapping("/user")
public class UserController {

    @Autowired
    private UserService userService;

    @PostMapping("register")
    public BaseResponse<Long> userRegister(@RequestBody UserRegisterRequest userRegisterRequest) {
        ThrowUtils.throwIf(ObjectUtil.isEmpty(userRegisterRequest), ErrorCode.PARAMS_ERROR);
        return ResultUtils.success(userService.userRegister(userRegisterRequest));
    }

    @PostMapping("login")
    public BaseResponse<LoginUserVO> userLogin(@RequestBody UserLoginRequest userLoginRequest, HttpServletRequest request) {
        ThrowUtils.throwIf(ObjectUtil.isEmpty(userLoginRequest), ErrorCode.PARAMS_ERROR);
        User user = userService.userLogin(userLoginRequest, request);
        return ResultUtils.success(userService.getUserVO(user));
    }

    @GetMapping("getLoginUser")
    public BaseResponse<LoginUserVO> getLoginUser(HttpServletRequest request) {
        User user = userService.getLoginUser(request);
        return ResultUtils.success(userService.getUserVO(user));
    }

    @PostMapping("logout")
    public BaseResponse<Boolean> userLogout(HttpServletRequest request) {
        return ResultUtils.success(userService.userLogout(request));
    }

    /**
     * 保存用户。
     *
     * @param user 用户
     * @return {@code true} 保存成功，{@code false} 保存失败
     */
    @PostMapping("save")
    @AuthCheck(mustRole = "admin")
    public BaseResponse<Boolean> save(@RequestBody User user) {
        ThrowUtils.throwIf(ObjectUtil.isEmpty(user), ErrorCode.PARAMS_ERROR);
        // TODO 参数校验
        return ResultUtils.success(userService.save(user));
    }

    /**
     * 根据主键删除用户。
     *
     * @param id 主键
     * @return {@code true} 删除成功，{@code false} 删除失败
     */
    @DeleteMapping("remove/{id}")
    @AuthCheck(mustRole = "admin")
    public BaseResponse<Boolean> remove(@PathVariable Long id) {
        ThrowUtils.throwIf(ObjectUtil.isEmpty(id), ErrorCode.PARAMS_ERROR);
        return ResultUtils.success(userService.removeById(id));
    }

    /**
     * 根据主键更新用户。
     *
     * @param user 用户
     * @return {@code true} 更新成功，{@code false} 更新失败
     */
    @PutMapping("update")
    public BaseResponse<Boolean> update(@RequestBody User user) {
        ThrowUtils.throwIf(ObjectUtil.isEmpty(user), ErrorCode.PARAMS_ERROR);
        return ResultUtils.success(userService.updateById(user));
    }

    /**
     * 查询所有用户。
     *
     * @return 所有数据
     */
    @GetMapping("list")
    @AuthCheck(mustRole = "admin")
    public BaseResponse<List<LoginUserVO>> list() {
        List<User> list = userService.list();
        return ResultUtils.success(userService.getUserVoList(list));
    }

    /**
     * 根据主键获取用户。
     *
     * @param id 用户主键
     * @return 用户详情
     */
    @GetMapping("getInfo/{id}")
    public BaseResponse<LoginUserVO> getInfo(@PathVariable Long id) {
        ThrowUtils.throwIf(ObjectUtil.isEmpty(id), ErrorCode.PARAMS_ERROR);
        User user = userService.getById(id);
        return ResultUtils.success(userService.getUserVO(user));
    }

    /**
     * 分页查询用户。
     *
     * @param userListRequest userListRequest
     * @return 分页对象
     */
    @PostMapping("page")
    public BaseResponse<Page<LoginUserVO>>page(@RequestBody UserListRequest userListRequest) {
        Page<User> userPage = userService.page(Page.of(userListRequest.getCurrent(), userListRequest.getPageSize()), userService.getUserListQueryWrapper(userListRequest));
        List<User> records = userPage.getRecords();
        List<LoginUserVO> loginUserVOList = userService.getUserVoList(records);
        // 创建新的 Page<LoginUserVO> 对象
        Page<LoginUserVO> voPage = new Page<>(userPage.getPageNumber(), userPage.getPageSize(), userPage.getTotalRow());
        voPage.setRecords(loginUserVOList);
        return ResultUtils.success(voPage);
    }

}

File: front-web/src/api/userController.ts
// @ts-ignore
/* eslint-disable */
import request from '@/utils/request'

/** 根据主键获取用户 GET /user/getInfo/${param0} */
export async function getInfo(
  // 叠加生成的Param类型 (非body参数swagger默认没有生成对象)
  params: API.getInfoParams,
  options?: { [key: string]: any }
) {
  const { id: param0, ...queryParams } = params
  return request<API.BaseResponseLoginUserVO>(`/user/getInfo/${param0}`, {
    method: 'GET',
    params: { ...queryParams },
    ...(options || {}),
  })
}

/** 获取当前登录用户 GET /user/getLoginUser */
export async function getLoginUser(options?: { [key: string]: any }) {
  return request<API.BaseResponseLoginUserVO>('/user/getLoginUser', {
    method: 'GET',
    ...(options || {}),
  })
}

/** 查询所有用户 GET /user/list */
export async function list(options?: { [key: string]: any }) {
  return request<API.BaseResponseListLoginUserVO>('/user/list', {
    method: 'GET',
    ...(options || {}),
  })
}

/** 用户登录 POST /user/login */
export async function userLogin(body: API.UserLoginRequest, options?: { [key: string]: any }) {
  return request<API.BaseResponseLoginUserVO>('/user/login', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    data: body,
    ...(options || {}),
  })
}

/** 用户登出 POST /user/logout */
export async function userLogout(options?: { [key: string]: any }) {
  return request<API.BaseResponseBoolean>('/user/logout', {
    method: 'POST',
    ...(options || {}),
  })
}

/** 分页查询用户 POST /user/page */
export async function page(body: API.UserListRequest, options?: { [key: string]: any }) {
  return request<API.BaseResponsePageLoginUserVO>('/user/page', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    data: body,
    ...(options || {}),
  })
}

/** 用户注册 POST /user/register */
export async function userRegister(
  body: API.UserRegisterRequest,
  options?: { [key: string]: any }
) {
  return request<API.BaseResponseLong>('/user/register', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    data: body,
    ...(options || {}),
  })
}

/** 根据主键删除用户 DELETE /user/remove/${param0} */
export async function remove(
  // 叠加生成的Param类型 (非body参数swagger默认没有生成对象)
  params: API.removeParams,
  options?: { [key: string]: any }
) {
  const { id: param0, ...queryParams } = params
  return request<API.BaseResponseBoolean>(`/user/remove/${param0}`, {
    method: 'DELETE',
    params: { ...queryParams },
    ...(options || {}),
  })
}

/** 保存用户 POST /user/save */
export async function save(body: API.User, options?: { [key: string]: any }) {
  return request<API.BaseResponseBoolean>('/user/save', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    data: body,
    ...(options || {}),
  })
}

/** 更新用户 PUT /user/update */
export async function update(body: API.User, options?: { [key: string]: any }) {
  return request<API.BaseResponseBoolean>('/user/update', {
    method: 'PUT',
    headers: {
      'Content-Type': 'application/json',
    },
    data: body,
    ...(options || {}),
  })
}
